research/fed-sm/job_configs/fedsm_prostate/app/custom/widgets/intime_model_fedsm_selector.py

```diff
@@ -0,0 +1,175 @@
+    def __init__(self, weigh_by_local_iter=False, aggregation_weights=None):
+        """Handler to determine if the model is globally best.
+        Note that only model_global and model_select participate in this selection process, while personalized models do not
+        Args:
+            weigh_by_local_iter (bool, optional): whether the metrics should be weighted by trainer's iteration number. Defaults to False.
+            aggregation_weights (dict, optional): a mapping of client name to float for aggregation. Defaults to None.
+        """
+        super().__init__()
+
+        self.val_metric_global = self.best_val_metric_global = -np.inf
+        self.val_metric_select = self.best_val_metric_select = -np.inf
+        self.weigh_by_local_iter = weigh_by_local_iter
+        self.validation_metric_name = MetaKey.INITIAL_METRICS
+        self.aggregation_weights = aggregation_weights or {}
+
+        self.logger.debug(f"model selection weights control: {aggregation_weights}")
+        self._reset_stats()
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        """Perform the handler process based on the event_type.
+
+        Args:
+            event_type (str): event type delivered from workflow
+            fl_ctx (FLContext): FL context, including peer context and other information
+        """
+        if event_type == EventType.START_RUN:
+            self._startup(fl_ctx)
+        elif event_type == EventType.BEFORE_PROCESS_SUBMISSION:
+            self._before_accept(fl_ctx)
+        elif event_type == AppEventType.BEFORE_AGGREGATION:
+            self._before_aggregate(fl_ctx)
+
+    def _startup(self, fl_ctx):
+        self._reset_stats()
+
+    def _reset_stats(self):
+        self.validation_mertic_global_weighted_sum = 0
+        self.validation_metric_global_sum_of_weights = 0
+        self.validation_mertic_select_weighted_sum = 0
+        self.validation_metric_select_sum_of_weights = 0
+
+    def _before_accept(self, fl_ctx: FLContext):
+        peer_ctx = fl_ctx.get_peer_context()
+        shareable: Shareable = peer_ctx.get_prop(FLContextKey.SHAREABLE)
+        try:
+            dxo = from_shareable(shareable)
+        except:
+            self.log_exception(fl_ctx, "shareable data is not a valid DXO")
+            return False
+
+        # check data_kind
+        if dxo.data_kind not in (
+            DataKind.WEIGHT_DIFF,
+            DataKind.WEIGHTS,
+            DataKind.COLLECTION,
+        ):
+            self.log_debug(fl_ctx, "I cannot handle {}".format(dxo.data_kind))
+            return False
+
+        if dxo.data is None:
+            self.log_debug(fl_ctx, "no data to filter")
+            return False
+
+        # DXO for FedSM is in "collection" format, containing three dxo objects (global_weights, person_weights, select_weights)
+        # together with the meta information
+        contribution_round = shareable.get_header(AppConstants.CONTRIBUTION_ROUND)
```

Comment:
This has been updated in this PR https://github.com/NVIDIA/NVFlare/pull/1401. I think we need the same changes here. CC @YuanTingHsieh

Reply:
yes, we need to update this part.

Reply:
@ZiyueXu77 then the minimum required version should be 2.3.0 I guess.

Reply:
will update this part